sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collationExpressions.scala

```diff
@@ -73,10 +74,9 @@ object CollateExpressionBuilder extends ExpressionBuilder {
  * This function is pass-through, it will not modify the input data.
  * Only type metadata will be updated.
  */
-case class Collate(child: Expression, collationName: String)
+case class Collate(child: Expression, collation: Either[UnresolvedCollation, ResolvedCollation])
```

Comment:
let's just do `collation: LogicalPlan`, then it's part of `Collate#children` and the new rule can just match `case UnresolvedCollation ` to resolve it.

Reply:
As discussed off-github, made `collation` as `Expression`, since both `ResolvedCollation` and `UnresolvedCollation` are Expressions.